### PR TITLE
Fix NPE in AbstractTarantoolConnectionManager

### LINKED ISSUE #87: If an address provider returns null, it produces NPE when establishing connections
Although this exception is then wrapped in the TarantoolConnectionException, there should be more clear exception anyway rather then a rough NPE:

```
Caused by: io.tarantool.driver.exceptions.TarantoolConnectionException: The client is not connected to Tarantool server
	at io.tarantool.driver.core.AbstractTarantoolConnectionManager.lambda$getConnection$0(AbstractTarantoolConnectionManager.java:96) ~[cartridge-driver-0.4.1.jar!/:na]
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.uniHandleStage(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.handle(Unknown Source) ~[na:na]
	at io.tarantool.driver.core.AbstractTarantoolConnectionManager.getConnection(AbstractTarantoolConnectionManager.java:88) ~[cartridge-driver-0.4.1.jar!/:na]
	at io.tarantool.driver.AbstractTarantoolClient.makeRequest(AbstractTarantoolClient.java:411) ~[cartridge-driver-0.4.1.jar!/:na]
	at io.tarantool.driver.AbstractTarantoolClient.call(AbstractTarantoolClient.java:228) ~[cartridge-driver-0.4.1.jar!/:na]
	at io.tarantool.driver.AbstractTarantoolClient.call(AbstractTarantoolClient.java:222) ~[cartridge-driver-0.4.1.jar!/:na]
	at io.tarantool.driver.AbstractTarantoolClient.call(AbstractTarantoolClient.java:216) ~[cartridge-driver-0.4.1.jar!/:na]
	... 33 common frames omitted
Caused by: java.lang.NullPointerException: null
	at io.tarantool.driver.core.AbstractTarantoolConnectionManager.getConnections(AbstractTarantoolConnectionManager.java:175) ~[cartridge-driver-0.4.1.jar!/:na]
	at io.tarantool.driver.core.AbstractTarantoolConnectionManager.establishConnections(AbstractTarantoolConnectionManager.java:160) ~[cartridge-driver-0.4.1.jar!/:na]
	at io.tarantool.driver.core.AbstractTarantoolConnectionManager.getConnectionInternal(AbstractTarantoolConnectionManager.java:112) ~[cartridge-driver-0.4.1.jar!/:na]
	... 38 common frames omitted
```

### PR DESCRIPTION
**Pull request by dkasimovskiy**
Return empty list instead of NPE in AbstractTarantoolConnectionManager.getConnections
Closes #87

### CODE REVIEW

**Review by vrogach2020 — CHANGES_REQUESTED** [review]

File: src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java [hunk 1]
```diff
@@ -175,8 +175,13 @@ public boolean refresh() {
         return result;
     }
 
-    private List<CompletableFuture<Map.Entry<TarantoolServerAddress, List<TarantoolConnection>>>> getConnections() {
+    private List<CompletableFuture<Map.Entry<TarantoolServerAddress, List<TarantoolConnection>>>> getConnections()
+            throws IllegalArgumentException {
         Collection<TarantoolServerAddress> addresses = getAddresses();
+        if (addresses == null) {
+            throw new IllegalArgumentException("TarantoolServerAddress collection couldn't be null");
```
[the remark this answers is not in the document]
Comment (dkasimovskiy) [reply src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java]: Done

File: src/main/java/io/tarantool/driver/core/connection/TarantoolClusterConnectionManager.java [hunk 2]
```diff
@@ -34,7 +36,7 @@ public TarantoolClusterConnectionManager(TarantoolClientConfig config,
 
     @Override
     protected Collection<TarantoolServerAddress> getAddresses() {
-        return addressProvider.getAddresses();
+        return Optional.ofNullable(addressProvider.getAddresses()).orElse(new ArrayList<>());
```
[the remark this answers is not in the document]
Comment (dkasimovskiy) [reply src/main/java/io/tarantool/driver/core/connection/TarantoolClusterConnectionManager.java]: Done

File: src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java [hunk 3]
```diff
@@ -175,8 +175,14 @@ public boolean refresh() {
         return result;
     }
 
-    private List<CompletableFuture<Map.Entry<TarantoolServerAddress, List<TarantoolConnection>>>> getConnections() {
+    private List<CompletableFuture<Map.Entry<TarantoolServerAddress, List<TarantoolConnection>>>> getConnections()
+            throws IllegalArgumentException {
         Collection<TarantoolServerAddress> addresses = getAddresses();
+        if (addresses == null) {
+            addresses = Collections.emptyList();
+            logger.warn("The list of server addresses is not defined");
```
Comment (vrogach2020) [inline src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java:183]: Логать тут можно, но только в debug.
Если возникнет такая ситуация под нагрузкой то warn забьёт всё место на сервере

File: src/test/java/io/tarantool/driver/integration/ConnectionIT.java [hunk 4]
```diff
@@ -322,4 +323,21 @@ public void testClientClosing_clientWithoutConnectionsShouldNotHang() throws Exc
...
+        TarantoolClientConfig config = TarantoolClientConfig.builder()
+                .withCredentials(new SimpleTarantoolCredentials(GUEST_USER, ""))
+                .withConnectTimeout(1000)
+                .withReadTimeout(1000)
+                .withConnections(1)
+                .build();
+
+        TarantoolClusterAddressProvider addressProvider = () -> null;
+
+        assertThrows(TarantoolClientException.class, () -> {
```
Comment (wey1and) [inline src/test/java/io/tarantool/driver/integration/ConnectionIT.java:338]: This assert doesn't check NPE

**Review by wey1and — CHANGES_REQUESTED** [review]

File: src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java [hunk 5]
```diff
@@ -175,8 +175,14 @@ public boolean refresh() {
...
+        }
+
```
[the remark this answers is not in the document]
Comment (vrogach2020) [reply src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java:185]: Why ? This method works correctly with empty lists. So we just cast null to empty list here

[the remark this answers is not in the document]
Comment (wey1and) [reply src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java:185]: I consider this the responsibility of the concrete implementation, not the abstract class.

[the remark this answers is not in the document]
Comment (vrogach2020) [reply src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java:185]: @dkasimovskiy discussed with @wey1and  - the decision is up to you :)  - keep the check here or move to concrete class - it doesn't matter. We are going to refactor this code a bit later.

File: src/test/java/io/tarantool/driver/integration/ConnectionIT.java (hunk 4 above)
Comment (dkasimovskiy) [reply src/test/java/io/tarantool/driver/integration/ConnectionIT.java:338]: Why do we need to check NPE if we are fix it?

Comment (wey1and) [reply src/test/java/io/tarantool/driver/integration/ConnectionIT.java:338]: We need to check that we don't have NPE anymore

File: src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dkasimovskiy) [reply src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java]: Let's keep it in the current place

File: src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java [hunk 6]
```diff
@@ -175,8 +175,14 @@ public boolean refresh() {
         return result;
     }
 
-    private List<CompletableFuture<Map.Entry<TarantoolServerAddress, List<TarantoolConnection>>>> getConnections() {
+    private List<CompletableFuture<Map.Entry<TarantoolServerAddress, List<TarantoolConnection>>>> getConnections()
+            throws IllegalArgumentException {
```
Comment (wey1and) [inline src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java:179]: IllegalArgumentException is redundant here

**Review by wey1and — CHANGES_REQUESTED** [review]

File: src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java (hunk 6 above)
Comment (dkasimovskiy) [reply src/main/java/io/tarantool/driver/core/connection/AbstractTarantoolConnectionManager.java]: Removed

**Review by wey1and — APPROVED** [review]

**Review by vrogach2020 — APPROVED** [review]

**Pull request merged**